File: windows/SakuraMusic/SakuraArt.cs
using System.Windows;
using System.Windows.Media;

namespace SakuraMusic;

/// <summary>Deterministic SplitMix64 so the branch is identical across launches.</summary>
public sealed class SeededRandom
{
    private ulong _state;
    public SeededRandom(ulong seed) => _state = seed + 0x9E3779B97F4A7C15UL;

    public ulong NextULong()
    {
        _state += 0x9E3779B97F4A7C15UL;
        ulong z = _state;
        z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9UL;
        z = (z ^ (z >> 27)) * 0x94D049BB133111EBUL;
        return z ^ (z >> 31);
    }

    public double NextDouble() => (NextULong() >> 11) * (1.0 / 9007199254740992.0);
    public double Range(double min, double max) => min + (max - min) * NextDouble();
    public bool Chance(double p) => NextDouble() < p;
    public T Pick<T>(IReadOnlyList<T> items) => items[(int)(NextULong() % (ulong)items.Count)];
}

public static class SakuraPalette
{
    public static Color Hex(uint v, double alpha = 1) =>
        Color.FromArgb((byte)Math.Round(alpha * 255), (byte)(v >> 16), (byte)(v >> 8), (byte)v);

    public static readonly Color PetalBlush = Hex(0xFFD6E4);
    public static readonly Color PetalPink = Hex(0xFFB7CE);
    public static readonly Color PetalRose = Hex(0xF799B8);
    public static readonly Color PetalWhite = Hex(0xFFF1F5);

    public static readonly Color[] FlowerFills = { Hex(0xFFC5D8), Hex(0xFFB3CB), Hex(0xF9A2BE), Hex(0xFFD9E6) };
    public static readonly Color FlowerEdge = Hex(0xE9779C, 0.55);
    public static readonly Color FlowerCenter = Hex(0xF7DE7A);
    public static readonly Color Bud = Hex(0xE8749A);
    public static readonly Color BranchWood = Hex(0x3D2523, 0.92);
}

public static class SakuraArt
{
    /// <summary>
    /// Petal outline in a unit box: base at (0.5, 1), notched tip at the top
    /// (WPF's y axis points down, so the tip is at y = 0).
    /// </summary>
    public static PathGeometry PetalGeometry(double width, double height, double offsetX = 0, double offsetY = 0)
    {
        Point P(double x, double y) => new(offsetX + x * width, offsetY + (1 - y) * height);

        var figure = new PathFigure { StartPoint = P(0.5, 0.0), IsClosed = true, IsFilled = true };
        figure.Segments.Add(new BezierSegment(P(0.02, 0.22), P(-0.02, 0.86), P(0.30, 0.95), true));
        figure.Segments.Add(new BezierSegment(P(0.41, 1.02), P(0.47, 0.86), P(0.5, 0.78), true));
        figure.Segments.Add(new BezierSegment(P(0.53, 0.86), P(0.59, 1.02), P(0.70, 0.95), true));
        figure.Segments.Add(new BezierSegment(P(1.02, 0.86), P(0.98, 0.22), P(0.5, 0.0), true));

        var geometry = new PathGeometry();
        geometry.Figures.Add(figure);
        return geometry;
    }

    /// <summary>Five-petal blossom centred on the origin.</summary>
    public static Geometry FlowerGeometry(double radius)
    {
        var group = new GeometryGroup { FillRule = FillRule.Nonzero };
        for (int i = 0; i < 5; i++)
        {
            var petal = PetalGeometry(radius * 0.84, radius, -radius * 0.42, -radius);
            petal.Transform = new RotateTransform(i * 72);
            group.Children.Add(petal);
        }
        return group;
    }

    /// <summary>Builds the corner branch as a frozen drawing, top-right aligned within <paramref name="size"/>.</summary>
    public static DrawingGroup MakeBranch(Size size, ulong seed = 20240401)
    {
        var rng = new SeededRandom(seed);
        var wood = new List<(PathGeometry path, double width)>();
        var flowers = new List<(Point center, double radius, Color fill, double rotation)>();
        var buds = new List<(Point center, double radius)>();

        // Work in a y-up space and flip at the end so the algorithm matches the macOS build.
        void Grow(Point start, double angle, double length, double width, int depth)
        {
            var end = new Point(start.X + Math.Cos(angle) * length, start.Y + Math.Sin(angle) * length);
            var normal = new Vector(-Math.Sin(angle), Math.Cos(angle));
            double bend = rng.Range(-0.28, 0.28) * length;
            var c1 = new Point(start.X + Math.Cos(angle) * length * 0.33 + normal.X * bend,
                               start.Y + Math.Sin(angle) * length * 0.33 + normal.Y * bend);
            var c2 = new Point(start.X + Math.Cos(angle) * length * 0.66 + normal.X * bend * 0.6,
                               start.Y + Math.Sin(angle) * length * 0.66 + normal.Y * bend * 0.6);

            var figure = new PathFigure { StartPoint = start, IsClosed = false, IsFilled = false };
            figure.Segments.Add(new BezierSegment(c1, c2, end, true));
            var path = new PathGeometry();
            path.Figures.Add(figure);
            wood.Add((path, width));

            if (depth >= 1)
            {
                int count = (int)(length / 22) + 1;
                for (int i = 0; i < count; i++)
                {
                    double t = rng.Range(0.15, 1.0);
                    var along = new Point(start.X + (end.X - start.X) * t, start.Y + (end.Y - start.Y) * t);
                    double side = rng.Range(-9, 9);
                    var center = new Point(along.X + normal.X * side, along.Y + normal.Y * side);
                    if (rng.Chance(0.22))
                        buds.Add((center, rng.Range(3.2, 4.6)));
                    else
                        flowers.Add((center, rng.Range(7.5, 14.5), rng.Pick(SakuraPalette.FlowerFills), rng.Range(0, 360)));
                }
            }

            if (depth >= 4 || length <= 18)
            {
                flowers.Add((end, rng.Range(9, 13), rng.Pick(SakuraPalette.FlowerFills), rng.Range(0, 360)));
                return;
            }

            Grow(end, angle + rng.Range(-0.32, 0.22), length * 0.74, width * 0.68, depth + 1);

            int shoots = depth == 0 ? 2 : (rng.Chance(0.75) ? 1 : 0);
            for (int i = 0; i < shoots; i++)
            {
                double sign = rng.Chance(0.5) ? 1 : -1;
                double t = rng.Range(0.35, 0.85);
                var origin = new Point(start.X + (end.X - start.X) * t, start.Y + (end.Y - start.Y) * t);
                Grow(origin, angle + sign * rng.Range(0.55, 1.05), length * 0.58, width * 0.5, depth + 1);
            }
        }

        var startPoint = new Point(size.Width + 14, size.Height - 26);
        Grow(startPoint, Math.PI + 0.16, size.Width * 0.34, 7.5, 0);

        var group = new DrawingGroup();
        // Flip y-up drawing space into WPF's y-down space.
        group.Transform = new ScaleTransform(1, -1, 0, size.Height / 2);

        var woodBrush = new SolidColorBrush(SakuraPalette.BranchWood);
        woodBrush.Freeze();
        foreach (var (path, width) in wood.OrderByDescending(w => w.width))
        {
            var pen = new Pen(woodBrush, Math.Max(1.2, width))
            {
                StartLineCap = PenLineCap.Round,
                EndLineCap = PenLineCap.Round,
                LineJoin = PenLineJoin.Round,
            };
            pen.Freeze();
            group.Children.Add(new GeometryDrawing(null, pen, path));
        }

        var budBrush = new SolidColorBrush(SakuraPalette.Bud);
        budBrush.Freeze();
        foreach (var (center, radius) in buds)
            group.Children.Add(new GeometryDrawing(budBrush, null, new EllipseGeometry(center, radius, radius)));

        var edgePen = new Pen(new SolidColorBrush(SakuraPalette.FlowerEdge), 0.6);
        edgePen.Freeze();
        var centerBrush = new SolidColorBrush(SakuraPalette.FlowerCenter);
        centerBrush.Freeze();

        foreach (var (center, radius, fill, rotation) in flowers)
        {
            var geometry = FlowerGeometry(radius);
            var transform = new TransformGroup();
            transform.Children.Add(new RotateTransform(rotation));
            transform.Children.Add(new TranslateTransform(center.X, center.Y));
            geometry.Transform = transform;
            var brush = new SolidColorBrush(fill);
            brush.Freeze();
            group.Children.Add(new GeometryDrawing(brush, edgePen, geometry));

            double r = radius * 0.18;
            group.Children.Add(new GeometryDrawing(centerBrush, null, new EllipseGeometry(center, r, r)));
        }

        group.Freeze();
        return group;
    }
}

File: macos/Sources/SakuraMusic/SakuraArt.swift
import AppKit
import QuartzCore

/// Deterministic random source so the branch looks the same on every launch.
struct SeededGenerator: RandomNumberGenerator {
    private var state: UInt64

    init(seed: UInt64) { state = seed &+ 0x9E37_79B9_7F4A_7C15 }

    mutating func next() -> UInt64 {
        state &+= 0x9E37_79B9_7F4A_7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58_476D_1CE4_E5B9
        z = (z ^ (z >> 27)) &* 0x94D0_49BB_1331_11EB
        return z ^ (z >> 31)
    }
}

enum SakuraPalette {
    static let space = CGColorSpace(name: CGColorSpace.sRGB)!

    static func rgba(_ r: CGFloat, _ g: CGFloat, _ b: CGFloat, _ a: CGFloat = 1) -> CGColor {
        CGColor(colorSpace: space, components: [r, g, b, a])!
    }

    static func hex(_ value: UInt32, alpha: CGFloat = 1) -> CGColor {
        rgba(
            CGFloat((value >> 16) & 0xFF) / 255,
            CGFloat((value >> 8) & 0xFF) / 255,
            CGFloat(value & 0xFF) / 255,
            alpha
        )
    }

    // Petals: pale to saturated sakura pinks.
    static let petalBlush = hex(0xFFD6E4)
    static let petalPink = hex(0xFFB7CE)
    static let petalRose = hex(0xF799B8)
    static let petalWhite = hex(0xFFF1F5)

    // Flowers on the branch.
    static let flowerFills: [CGColor] = [hex(0xFFC5D8), hex(0xFFB3CB), hex(0xF9A2BE), hex(0xFFD9E6)]
    static let flowerEdge = hex(0xE9779C, alpha: 0.55)
    static let flowerCenter = hex(0xF7DE7A)
    static let bud = hex(0xE8749A)

    static let branchWood = hex(0x3D2523, alpha: 0.92)
}

enum SakuraArt {
    /// Sakura petal outline: a teardrop with a notched tip, base at the bottom
    /// centre of `rect`, tip at the top.
    static func petalPath(in rect: CGRect) -> CGPath {
        func pt(_ x: CGFloat, _ y: CGFloat) -> CGPoint {
            CGPoint(x: rect.minX + x * rect.width, y: rect.minY + y * rect.height)
        }
        let p = CGMutablePath()
        p.move(to: pt(0.5, 0.0))
        p.addCurve(to: pt(0.30, 0.95), control1: pt(0.02, 0.22), control2: pt(-0.02, 0.86))
        p.addCurve(to: pt(0.5, 0.78), control1: pt(0.41, 1.02), control2: pt(0.47, 0.86))
        p.addCurve(to: pt(0.70, 0.95), control1: pt(0.53, 0.86), control2: pt(0.59, 1.02))
        p.addCurve(to: pt(0.5, 0.0), control1: pt(1.02, 0.86), control2: pt(0.98, 0.22))
        p.closeSubpath()
        return p
    }

    /// Five-petal blossom centred at the origin.
    static func flowerPath(radius: CGFloat) -> CGPath {
        let p = CGMutablePath()
        let petalRect = CGRect(x: -radius * 0.42, y: 0, width: radius * 0.84, height: radius)
        let petal = petalPath(in: petalRect)
        for i in 0..<5 {
            let rotation = CGAffineTransform(rotationAngle: CGFloat(i) * (2 * .pi / 5))
            p.addPath(petal, transform: rotation)
        }
        return p
    }

    /// Renders a soft-shaded petal sprite for the particle emitter.
    static func petalImage(size: CGFloat, scale: CGFloat, base: CGColor, tip: CGColor) -> CGImage? {
        let px = Int(size * scale)
        guard let ctx = CGContext(
            data: nil, width: px, height: px, bitsPerComponent: 8, bytesPerRow: 0,
            space: SakuraPalette.space, bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
        ) else { return nil }

        ctx.scaleBy(x: scale, y: scale)
        ctx.setShouldAntialias(true)

        let inset: CGFloat = 1.5
        let rect = CGRect(x: inset, y: inset, width: size - inset * 2, height: size - inset * 2)
        ctx.saveGState()
        ctx.addPath(petalPath(in: rect))
        ctx.clip()
        if let gradient = CGGradient(
            colorsSpace: SakuraPalette.space, colors: [base, tip] as CFArray, locations: [0, 1]
        ) {
            ctx.drawLinearGradient(
                gradient,
                start: CGPoint(x: size / 2, y: rect.minY),
                end: CGPoint(x: size / 2, y: rect.maxY),
                options: []
            )
        }
        ctx.restoreGState()

        // Faint central vein for a little depth.
        ctx.setStrokeColor(SakuraPalette.rgba(1, 1, 1, 0.35))
        ctx.setLineWidth(0.8)
        ctx.move(to: CGPoint(x: size / 2, y: rect.minY + rect.height * 0.08))
        ctx.addLine(to: CGPoint(x: size / 2, y: rect.minY + rect.height * 0.62))
        ctx.strokePath()

        return ctx.makeImage()
    }

    // MARK: - Branch

    /// Builds a layer tree with a cherry branch that reaches in from the top-right
    /// corner of a `size`-sized area. The branch is generated once from a fixed
    /// seed so it is identical across windows and launches.
    static func makeBranchLayer(size: CGSize, seed: UInt64 = 20240401) -> CALayer {
        let root = CALayer()
        root.bounds = CGRect(origin: .zero, size: size)
        root.anchorPoint = CGPoint(x: 1, y: 1)
        root.masksToBounds = false

        var rng = SeededGenerator(seed: seed)
        var segments: [(path: CGPath, width: CGFloat)] = []
        var flowers: [(center: CGPoint, radius: CGFloat, fill: CGColor, rotation: CGFloat)] = []
        var buds: [(center: CGPoint, radius: CGFloat)] = []

        func grow(from start: CGPoint, angle: CGFloat, length: CGFloat, width: CGFloat, depth: Int) {
            let end = CGPoint(x: start.x + cos(angle) * length, y: start.y + sin(angle) * length)
            let normal = CGPoint(x: -sin(angle), y: cos(angle))
            let bend = CGFloat.random(in: -0.28...0.28, using: &rng) * length
            let c1 = CGPoint(
                x: start.x + cos(angle) * length * 0.33 + normal.x * bend,
                y: start.y + sin(angle) * length * 0.33 + normal.y * bend
            )
            let c2 = CGPoint(
                x: start.x + cos(angle) * length * 0.66 + normal.x * bend * 0.6,
                y: start.y + sin(angle) * length * 0.66 + normal.y * bend * 0.6
            )
            let path = CGMutablePath()
            path.move(to: start)
            path.addCurve(to: end, control1: c1, control2: c2)
            segments.append((path, width))

            // Blossoms cluster on the thinner growth.
            if depth >= 1 {
                let count = Int(length / 22) + 1
                for _ in 0..<count {
                    let t = CGFloat.random(in: 0.15...1.0, using: &rng)
                    let along = CGPoint(x: start.x + (end.x - start.x) * t, y: start.y + (end.y - start.y) * t)
                    let side = CGFloat.random(in: -9...9, using: &rng)
                    let center = CGPoint(x: along.x + normal.x * side, y: along.y + normal.y * side)
                    if Double.random(in: 0...1, using: &rng) < 0.22 {
                        buds.append((center, CGFloat.random(in: 3.2...4.6, using: &rng)))
                    } else {
                        flowers.append((
                            center,
                            CGFloat.random(in: 7.5...14.5, using: &rng),
                            SakuraPalette.flowerFills.randomElement(using: &rng)!,
                            CGFloat.random(in: 0...(2 * .pi), using: &rng)
                        ))
                    }
                }
            }

            guard depth < 4, length > 18 else {
                flowers.append((end, CGFloat.random(in: 9...13, using: &rng),
                                SakuraPalette.flowerFills.randomElement(using: &rng)!,
                                CGFloat.random(in: 0...(2 * .pi), using: &rng)))
                return
            }

            // Main continuation.
            grow(from: end, angle: angle + CGFloat.random(in: -0.32...0.22, using: &rng),
                 length: length * 0.74, width: width * 0.68, depth: depth + 1)

            // Side shoots.
            let shoots = depth == 0 ? 2 : (Double.random(in: 0...1, using: &rng) < 0.75 ? 1 : 0)
            for _ in 0..<shoots {
                let sign: CGFloat = Bool.random(using: &rng) ? 1 : -1
                let t = CGFloat.random(in: 0.35...0.85, using: &rng)
                let origin = CGPoint(x: start.x + (end.x - start.x) * t, y: start.y + (end.y - start.y) * t)
                grow(from: origin, angle: angle + sign * CGFloat.random(in: 0.55...1.05, using: &rng),
                     length: length * 0.58, width: width * 0.5, depth: depth + 1)
            }
        }

        // Enter from just outside the top-right corner, heading left and slightly down.
        let start = CGPoint(x: size.width + 14, y: size.height - 26)
        grow(from: start, angle: .pi + 0.16, length: size.width * 0.34, width: 7.5, depth: 0)

        // Wood, thick to thin so thin segments sit on top of thick ones.
        for segment in segments.sorted(by: { $0.width > $1.width }) {
            let shape = CAShapeLayer()
            shape.path = segment.path
            shape.strokeColor = SakuraPalette.branchWood
            shape.fillColor = nil
            shape.lineWidth = max(1.2, segment.width)
            shape.lineCap = .round
            shape.lineJoin = .round
            root.addSublayer(shape)
        }

        for bud in buds {
            let shape = CAShapeLayer()
            shape.path = CGPath(ellipseIn: CGRect(x: -bud.radius, y: -bud.radius,
                                                  width: bud.radius * 2, height: bud.radius * 2), transform: nil)
            shape.fillColor = SakuraPalette.bud
            shape.position = bud.center
            root.addSublayer(shape)
        }

        for flower in flowers {
            let petals = CAShapeLayer()
            petals.path = flowerPath(radius: flower.radius)
            petals.fillColor = flower.fill
            petals.strokeColor = SakuraPalette.flowerEdge
            petals.lineWidth = 0.6
            petals.position = flower.center
            petals.setAffineTransform(CGAffineTransform(rotationAngle: flower.rotation))
            root.addSublayer(petals)

            let center = CAShapeLayer()
            let r = flower.radius * 0.18
            center.path = CGPath(ellipseIn: CGRect(x: -r, y: -r, width: r * 2, height: r * 2), transform: nil)
            center.fillColor = SakuraPalette.flowerCenter
            center.position = flower.center
            root.addSublayer(center)
        }

        root.shadowColor = SakuraPalette.rgba(0.25, 0.05, 0.1, 1)
        root.shadowOpacity = 0.28
        root.shadowRadius = 5
        root.shadowOffset = CGSize(width: 0, height: -2)
        root.shouldRasterize = true
        return root
    }
}
